File: 0648-replace-words/0648-replace-words.cpp
class Node{
    public:
    vector<Node*>arr;
    bool isEnd;
    Node(){isEnd=false;arr.resize(26,NULL);}
    bool isthere(char c){
        return (arr[c-'a']==NULL?false:true);
    }
    Node* give(char c){
        if(isthere(c)){
            return arr[c-'a'];
        }
        return NULL;
    }
    void put(Node* node,char c){
        arr[c-'a']=node;
    }
};
class Trie{
    public:
    Node* root;
    Trie(){
        root=new Node();
    }
    void insert(string s){
        Node* node=root;
        for(int i=0;i<s.size();i++){
            char c=s[i];
            if(node->isthere(c)){
                node=node->give(c);
            }
            else{
                Node* new_node=new Node();
                node->arr[c-'a']=new_node;
                node=new_node;
            }
        }
        node->isEnd=true;
    }
    string search(string s){
        Node* node=root;
        string ans="";
        for(int i=0;i<s.size();i++){
            char c=s[i];
            if(node->isthere(c)){
                node=node->give(c);
                ans+=c;
                if(node->isEnd){return ans;}
            }
            else{
              return s;
            }
        }
        return ans;

    }
};
class Solution {
public:
    string replaceWords(vector<string>& dic, string sen) {
        Trie* obj=new Trie();
        for(auto i:dic){
            obj->insert(i);
        }
        string ans="";
        vector<string>v;
        sen+=' ';
        string small="";
        for(auto i:sen){
            if(i==' '){
                v.push_back(small);
                small="";
                continue;
            }
            small+=i;
        }
        for(auto i:v){
            ans+=obj->search(i)+" ";
            
        }
        return ans.substr(0,ans.size()-1);
        
    }
};

File: 0648-replace-words/0648-replace-words.java

class Node {
    Node[] arr;
    boolean isEnd;

    Node() {
        isEnd = false;
        arr = new Node[26];
    }

    boolean isthere(char c) {
        return arr[c - 'a'] != null;
    }

    Node give(char c) {
        return isthere(c) ? arr[c - 'a'] : null;
    }

    void put(Node node, char c) {
        arr[c - 'a'] = node;
    }
}

class Trie {
    Node root;

    Trie() {
        root = new Node();
    }

    void insert(String s) {
        Node node = root;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (node.isthere(c)) {
                node = node.give(c);
            } else {
                Node new_node = new Node();
                node.put(new_node, c);
                node = new_node;
            }
        }
        node.isEnd = true;
    }

    String search(String s) {
        Node node = root;
        StringBuilder ans = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (node.isthere(c)) {
                node = node.give(c);
                ans.append(c);
                if (node.isEnd) {
                    return ans.toString();
                }
            } else {
                return s;
            }
        }
        return ans.toString();
    }
}

public class Solution {
    public String replaceWords(List<String> dic, String sen) {
        Trie obj = new Trie();
        for (String i : dic) {
            obj.insert(i);
        }
        StringBuilder ans = new StringBuilder();
        String[] words = sen.split(" ");
        for (String word : words) {
            ans.append(obj.search(word)).append(" ");
        }
        return ans.toString().trim();
    }
}